File: Modules/HumanResource/TrainingManagement/MasterProgram/MasterProgramApplyForm.cs
using Serenity.ComponentModel;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace HRMSoftware.TrainingManagement.Forms;

[FormScript("TrainingManagement.MasterProgramApply")]
[BasedOnRow(typeof(MasterProgramRow), CheckNames = true)]
public class MasterProgramApplyForm
{
    [Tab("Basic Information")]
    [Category("Basic Information")]
    public string ProgramName { get; set; }
    public List<int> DepartmentList { get; set; }
    [HalfWidth]
    public string Routine { get; set; }
    [HalfWidth]
    public int RoutineInterval { get; set; }
    [HalfWidth]
    public DateTime RoutineStartDate { get; set; }
    [HalfWidth]
    public DateTime RoutineEndDate { get; set; }
    public string Status { get; set; }
    [TextAreaEditor(Rows = 3)]
    public string Comment { get; set; }
    [Category("Detail Information")]
    [TextAreaEditor(Rows = 10)]
    public string Detail { get; set; }
    public string ExtraDocument { get; set; }
    [Tab("Program Flow")]
    [IgnoreName, LabelWidth("0"), MasterProgramFlowEditor]
    public List<ProgramFlowRow> FlowList { get; set; }
}
    // public int InsertUserId { get; set; }
    // public DateTime InsertDate { get; set; }
    // public int UpdateUserId { get; set; }
    // public DateTime UpdateDate { get; set; }
    // public int DeleteUserId { get; set; }
    // public DateTime DeleteDate { get; set; }

File: Modules/ServerTypes/TrainingManagement/MasterProgramApplyForm.ts
import { StringEditor, LookupEditor, EnumEditor, IntegerEditor, DateEditor, TextAreaEditor, MultipleImageUploadEditor, PrefixedContext } from "@serenity-is/corelib";
import { MasterProgramRoutineType } from "./MasterProgramRoutineType";
import { MasterProgramStatusType } from "./MasterProgramStatusType";
import { MasterProgramFlowEditor } from "@/HumanResource/TrainingManagement/ProgramFlow/MasterProgramFlowEditor";
import { initFormType } from "@serenity-is/corelib/q";

export interface MasterProgramApplyForm {
    ProgramName: StringEditor;
    DepartmentList: LookupEditor;
    Routine: EnumEditor;
    RoutineInterval: IntegerEditor;
    RoutineStartDate: DateEditor;
    RoutineEndDate: DateEditor;
    Status: EnumEditor;
    Comment: TextAreaEditor;
    Detail: TextAreaEditor;
    ExtraDocument: MultipleImageUploadEditor;
    FlowList: MasterProgramFlowEditor;
}

export class MasterProgramApplyForm extends PrefixedContext {
    static formKey = 'TrainingManagement.MasterProgramApply';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!MasterProgramApplyForm.init)  {
            MasterProgramApplyForm.init = true;

            var w0 = StringEditor;
            var w1 = LookupEditor;
            var w2 = EnumEditor;
            var w3 = IntegerEditor;
            var w4 = DateEditor;
            var w5 = TextAreaEditor;
            var w6 = MultipleImageUploadEditor;
            var w7 = MasterProgramFlowEditor;

            initFormType(MasterProgramApplyForm, [
                'ProgramName', w0,
                'DepartmentList', w1,
                'Routine', w2,
                'RoutineInterval', w3,
                'RoutineStartDate', w4,
                'RoutineEndDate', w4,
                'Status', w2,
                'Comment', w5,
                'Detail', w5,
                'ExtraDocument', w6,
                'FlowList', w7
            ]);
        }
    }
}

[MasterProgramRoutineType, MasterProgramStatusType]; // referenced types